### PR TITLE
8294809: ListenerHelper for managing and disconnecting listeners

### PR DESCRIPTION
**Pull request by andy-goryachev-oracle**
Introduction

There is a number of places where various listeners (strong as well as weak) are added, to be later disconnected in one go. For example, Skin implementations use dispose() method to clean up the listeners installed in the corresponding Control (sometimes using LambdaMultiplePropertyChangeListenerHandler class).

This pattern is also used by app developers, but there is no public utility class to do that, so every one must invent their own, see for instance
https://github.com/andy-goryachev/FxTextEditor/blob/master/src/goryachev/fx/FxDisconnector.java

Proposal

It might be beneficial to create a class (ListenerHelper, feel free to suggest a better name) which:

- provides convenient methods like addChangeListener, addInvalidationListener, addWeakChangeListener, etc.
- keeps track of the listeners and the corresponding ObservableValues
- provides a single disconnect() method to remove all the listeners in one go.
- optionally, it should be possible to add a lambda (Runnable) to a group of properties
- optionally, there should be a boolean flag to fire the lambda immediately
- strongly suggest implementing an IDisconnectable functional interface with a single disconnect() method

Make it Public Later

Initially, I think we could place the new class in package com.sun.javafx.scene.control; to be made publicly accessible later.

### CODE REVIEW

File: modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java [hunk 1]
```diff
@@ -0,0 +1,706 @@
...
+            throw new NullPointerException("onChange must not be null.");
+        }
+
+        ChLi li = new ChLi() {
+            @Override
+            public void disconnect() {
+                for (ObservableValue p : props) {
+                    p.removeListener(this);
+                }
+            }
```
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java:116]: I wonder if 
```
items.remove(this);
```
should be added here and in other places.

Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java:124]: added

**Comment by andy-goryachev-oracle** [comment]
Thank you @nlisker and @hjohn for providing the link and explanations.  

You are right, ListenerHelper is somewhat similar to Subscription in that it allows for quick disconnect().

It is, however, designed for a different purpose, that goes beyond what is needed from a simple listener management.  Specifically,

- it assumes multiple listeners/handlers/callbacks from the start
- it allows for invoking listeners/callbacks immediately as it is a frequent pattern not only in Skins but also in application code (I do want to make this class a part of public API eventually)
- unlike Subscription, we can addEventHandlers and addEventFilters
- unlike Subscription, we can add Weak listeners/callbacks, though I admit it might be a design decision to use or not to use that

to summarize, I see some minor overlap between Subscription and ListenerHelper, but mostly these are two different beasts designed for two different purposes.

I do agree with the call to cooperate!  Would love to.

what do you think?

**Comment by andy-goryachev-oracle** [comment]
> Reading your proposal I see you might be looking to support weak
listeners as well. Personally, I believe these are better replaced 

You have a *very* good point here, considering the purpose of this class.  Let me see if support for weak listeners should be removed. Thank you!

**Comment by andy-goryachev-oracle** [comment]
Support for weak listeners cannot be removed, for example in a situation when Control gets removed from the scene graph and therefore becomes eligible for collection.

File: modules/javafx.controls/src/main/java/javafx/scene/control/SkinBase.java [hunk 2]
```diff
@@ -82,8 +83,10 @@
      * want to adjust the way listeners are added rather than continuing to use this
      * map (although it doesn't really do much harm).
      */
+    @Deprecated // replace with listenerHelper
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/main/java/javafx/scene/control/SkinBase.java:86]: We only use `@Deprecated` annotation on public API, so you should limit this to a comment.

File: modules/javafx.controls/src/main/java/javafx/scene/control/SkinBase.java [hunk 3]
```diff
@@ -207,6 +214,17 @@ protected final void consumeMouseEvents(boolean value) {
         }
     }
 
+    /**
+     * Returns the skin's instance of {@link ListenerHelper}, creating it if necessary.
+     *
+     * @since 20
+     */
+    protected ListenerHelper listenerHelper() {
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/main/java/javafx/scene/control/SkinBase.java:222]: This would make it part of the public API, so you will need to find some other way to do this without adding anything to the public API surface.

File: modules/javafx.controls/src/main/java/javafx/scene/control/SkinBase.java [hunk 4]
```diff
@@ -219,6 +237,7 @@ protected final void consumeMouseEvents(boolean value) {
      *  may be {@code null}
      * @since 9
      */
+    // TODO I would like to deprecate and remove these methods, and replace them by listenerHelper().add**()
     protected final void registerChangeListener(ObservableValue<?> observable, Consumer<ObservableValue<?>> operation) {
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/main/java/javafx/scene/control/SkinBase.java:241]: I don't think we want to deprecate this. It is part of the public API, so there would need to be a compelling reason to deprecate it.

**Review by kevinrushforth — CHANGES_REQUESTED** [review]
The stated intention of this PR is to provide a helpful utility to enable Skins to more easily manage their listeners to avoid memory leaks. I am not sold on the idea of this migrating to public API, but we can defer that (longer) discussion. As long as this does not impact the public API surface, we can evaluate it as a internal helper class that enables a few memory leak bug fixes.

To that end, you will need at least one change, since this PR _does_ add to the public API surface (and exposes an internal type in the process).

File: modules/javafx.controls/src/main/java/javafx/scene/control/SkinBase.java (hunk 4 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/SkinBase.java:241]: we can keep it, eventually replacing the implementation with ListenerHelper, see [JDK-8296076](https://bugs.openjdk.org/browse/JDK-8296076)

File: modules/javafx.controls/src/main/java/javafx/scene/control/SkinBase.java (hunk 3 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/SkinBase.java:222]: I *do* want this as public API.

I can change the code to use an accessor, but it gets really ugly:

```
ListenerHelper.get(PaginationSkin.this).addListChangeListener(getSkinnable().getStyleClass(), (ch) -> {
```

Besides, what's the harm of adding a [better] public API?  I do want ListenerHelper to be public anyway.

Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/SkinBase.java]: changed the code to remove listenerHelper() to avoid changing the public API.
For now, will be using accessor, and using a local variable in skins where we have many calls to LH (in constructors, for example).

**Review by kevinrushforth — COMMENTED** [review]
Before reviewing the code itself, I have a couple global comments and a small number of inline comments.

Since `ListenerHelper` is now an internal implementation detail, we can defer the longer discussion of whether and how to expose this as public API. In its current incarnation, it seems like a helpful utility to fix a number of skin memory leaks, and will likely be an eventual replacement for the (also internal) `LambdaMultiplePropertyChangeListenerHandler` utility class.

If this ever does become public API in the future, we will need a naming discussion as part of the larger discussion of what form the API should take. For example, we wouldn't use `IDisconnectable` as the name of a public interface in the JavaFX API (we don't use the pattern a leading `I` for interfaces anywhere in JavaFX or in the JDK); also, we would need more descriptive names for `ChLi`, `MaChLi`, and so forth (presuming they need to be exposed).

The larger discussion, though, will first need to clearly identify what the purpose of this utility is. If it is specific to control skins, then whatever minimal public API surface that needs to be exposed to do that job so third-party controls could use it is what we would want. If it is to be more general, then it doesn't belong in the `javafx.controls` module at all, but belongs in `javafx.base`. In the either case, but especially for an API in `javafx.base`, we need to ensure that this is generally useful to application developers, and not just for our own internal UI controls. Also, we would need to see how it fits with the existing APIs in `javafx.base`, and the newly proposed APIs in PR #830. Btw, the existing implementation is not suitable to be generic, since it references classes like `Task`, `Node`, `Scene`, `Label`, and `TableColumn`.

As I mentioned above, we don't need to have this discussion unless and until there is a desire to make this public API some future release.

**Comment by kevinrushforth** [comment]
@arapte and @aghaisas Can you also review this? It really needs to be done in connection with one of the existing Draft PRs that will take advantage of it, such as PR #925 , since the purpose is to make it easier for skins to avoid memory leaks.

**Comment by andy-goryachev-oracle** [comment]
I fully agree with the need for a wider discussion.  Any suggestions for changing of names etc. are always welcomed.

The use case for a public ListenerHelper is to provide a convenient way to disconnect any listeners when a certain application-level UI (panel, window) is closed and gets discarded.  For instance, a financial application might have a short-lived popup with a chart, multiple subscriptions to real time data, and so on; and there is a single moment when this popup gets closed and discarded.  In this situation, it is better to disconnect all data subscriptions as well as all listeners that has been added to various application components, at that exact moment.

For this PR, the goal is not to extend the public API surface and keep things as implementation detail.  ChLi and other can be made package private or private.

File: modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/IDisconnectable.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * A functional interface that provides a {@link #disconnect()} method.
+ * <p>
+ * Original code is re-licensed to Oracle by the author.
+ * https://github.com/andy-goryachev/FxTextEditor/blob/master/src/goryachev/common/util/Disconnectable.java
+ * Copyright © 2021-2022 Andy Goryachev <andy@goryachev.com>
+ */
+@FunctionalInterface
+public interface IDisconnectable {
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/IDisconnectable.java:35]: leaving as is for now.

**Comment by andy-goryachev-oracle** [comment]
@arapte and @aghaisas : 
could you please take a look at this?  there is a whole bunch of dependent PRs that will get unblocked once this change is in.
Thank you

File: modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java [hunk 6]
```diff
@@ -0,0 +1,508 @@
...
+
+    public ListenerHelper() {
+    }
+
+    public static void setAccessor(Function<SkinBase<?>,ListenerHelper> a) {
+        accessor = a;
+    }
+
+    public static ListenerHelper get(SkinBase<?> skin) {
+        return accessor.apply(skin);
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java:87]: I don't think this is the case.
When this method is called, SkinBase class must be already loaded and resolved, that means it's static { } block, which sets the accessor, has been executed.  But just in case, I should move the static block in SkinBase, so we don't have any surprises.

File: modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+ * <p>
+ * There are two usage patterns:
+ * <ul>
+ * <li>Client code registers a number of listeners and removes them all at once
+ * via {@link #disconnect()} call.
+ * <li>Client code registers a number of listeners and removes one via its
+ * {@link IDisconnectable} instance.
+ * </ul>
+ */
+public class ListenerHelper implements IDisconnectable {
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java:70]: you are correct: the original intent for this class was to make it a general purpose facility to help with listeners, something that might be useful at the application level (and it used `ListenerHelper.get(Node)`).  Since that would require CSR and a public discussion, we decided to hide it as an implementation detail in skins (to unblock skin memory leak fixes).

Once we go through all the necessary discussions and everyone agrees, we could introduce it as a public API in the javafx.base module.

[the remark this answers is not in the document]
Comment (kevinrushforth) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java:70]: Exactly. In its current form, `ListenerHelper` is a utility helper class for Skins, and should be reviewed with that in mind. As discussed in previous comments, there are several things that will need to change when/if this is proposed as a more general utility. We can defer that discuss, since this is entirely an internal helper class for now.

As a next step, after this is integrated and before any discussion of making this a public utility, it can be used as a replacement for `LambdaMultiplePropertyChangeListenerHandler` -- see [JDK-8296076](https://bugs.openjdk.org/browse/JDK-8296076).

File: modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (aghaisas) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java:87]: The `accessor` cannot be null **only if** it is used from *Skins.
In the current form, as we provide a no-argument public constructor, it is possible to create an object of `ListenerHelper` without calling setAccessor. Many unit tests added as part of this PR create such `ListenerHelper` objects.

File: modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (aghaisas) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java:70]: Although it is an internal (non-public) utility class, once introduced, it will be tempting to use it from non-Skin code as well.
If the current scope is limited to provide "a utility helper class for Skins" then how about doing one of two things-
1) Mention the current intention and possible future modifications as a block comment above the `ListenerHelper` class
2) Keep only common methods in`ListenerHelper` class and move `Skin` specific code to a class derived from `ListenerHelper`. Skins should start using this derived class.

File: modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java:87]: The `accessor` is used **only** if used from *Skins, meaning to get it from a Skin, one needs to invoke `get(SkinBase)` method, and it's magically not a null there!

It is perfectly fine to create an instance of ListenerHelper using the constructor.  The `accessor` will be null, of course, but it's ok because it will never be used until one calls `get(SkinBase)` ... and see above.

[the remark this answers is not in the document]
Comment (aghaisas) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java:91]: That is my whole point. We have two paths - one for Skin classes and other for generic usage.

Approving the PR as you have added a comment regarding `ListenerHelper` class's current scope being limited to Skins. We may need to revisit the suggestion of having a derived class in future.

**Review by aghaisas — APPROVED** [review]

**Comment by kevinrushforth** [comment]
As part of reviewing this, I looked at a couple of the fixes that will make use of this, and I have a high-level question on the usage pattern.

Along with the needed changes to fix the leak, PR #925 replaces four calls to `registerChangeListener` -- which is part of the public `SkinBase` API for skins to use to register a listener that is cleaned up when the skin is disposed or when `unregisterChangeListeners` is called. The current implementation of register/unregister methods uses `LambdaMultiplePropertyChangeListenerHandler`, which `ListenerHelper` intends to replace, but I wonder whether it might be better for the individual skins continue to call `registerChangeListener` (along with the related `registerInvalidationListener` and `registerListChangeListener` methods), where those methods are currently used? Eventually, the implementation of `registerChangeListener` and friends can replace `LambdaMultiplePropertyChangeListenerHandler` with `ListenerHelper`. This would minimize the changes needed for the individual memory leak bugs.

So my main question is: is there anything about the addition of ListenerChangeHelper that will make skins that continue to use registerListenerHelper stop working? I suspect not, since I tested PR #925 with that part of the change reverted, and it still works without leaking. Given this, I wondered why you made that change in the `PaginationSkin` (and others for other PRs), and figured I might be missing something.

**Comment by andy-goryachev-oracle** [comment]
Kevin:

my initial intent was to label register*() methods as @Deprecated and switch completely to `ListenerHelper` (LH).  Since there is general reluctance to alter public APIs (quite understandable) the next bright idea is to keep the methods, but switch implementation to use LH, [JDK-8296076](https://bugs.openjdk.org/browse/JDK-8296076).

At the end, I'd like to have 0 callers of register*() inside javafx, keeping these methods for any external skins that might use it.  The main reason is orthogonality - LH has a richer set of methods, which register*() support just a few, and are themselves delegates.

**Comment by andy-goryachev-oracle** [comment]
> we would need more descriptive names for `ChLi`, `MaChLi`, and so forth (presuming they need to be exposed).

they don't need to be exposed, will convert them to `private`.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TestListenerHelper.java [hunk 8]
```diff
@@ -0,0 +1,465 @@
...
+        assertEquals(1, ct.get());
+
+        h.disconnect();
+
+        p.set("2");
+        assertEquals(1, ct.get());
+    }
+
+    @Test
+    public void testInvalidationListener_Callback_FireImmediately() {
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/TestListenerHelper.java:295]: Minor: redundant (tests the same method signature as `testInvalidationListener_Plain_FireImmediately`.

**Review by kevinrushforth — APPROVED** [review]
This looks ready to go in. I left a few minor comments and questions, but I am approving it as is.

**Review by aghaisas — APPROVED** [review]

File: modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/IDisconnectable.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+// Original code is re-licensed to Oracle by the author.
+// https://github.com/andy-goryachev/FxTextEditor/blob/master/src/goryachev/common/util/Disconnectable.java
+// Copyright © 2021-2022 Andy Goryachev <andy@goryachev.com>
+package com.sun.javafx.scene.control;
+
+/**
+ * A functional interface that provides a {@link #disconnect()} method.
+ */
+@FunctionalInterface
+public interface IDisconnectable {
```
Comment (Maran23) [inline modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/IDisconnectable.java:34]: I'm still in favor of remove the `I` in front of this interface. This is .NET convention and not used in Java (at least I never saw it, and thinking of naming `Iterable` -> `IIterable` would be weird anyway.

Comment (kevinrushforth) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/IDisconnectable.java:34]: I made a [similar comment](https://github.com/openjdk/jfx/pull/908#pullrequestreview-1168974400) about not using the `I` prefix earlier. Before this can be considered for public API (which won't be soon), this interface will need to be renamed. It seems best to discuss what it should be renamed to as part of the larger discussion about the API surface, functionality, and naming, so I didn't ask that it be renamed now.

Comment (Maran23) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/IDisconnectable.java:34]: Agreed.

Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/IDisconnectable.java:34]: any suggestions?  Disconnectable?   HasDisconnect?

File: modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java [hunk 10]
```diff
@@ -0,0 +1,512 @@
...
+ * This class is currently used for clean replacement of {@link Skin}s.
+ * We should consider making this class a part of the public API in {@code javax.base},
+ * since it proved itself useful in removing listeners and handlers in bulk at the application level.
+ */
+public class ListenerHelper implements IDisconnectable {
+    private WeakReference<Object> ownerRef;
+    private final ArrayList<IDisconnectable> items = new ArrayList<>(4);
+    private static Function<SkinBase<?>,ListenerHelper> accessor;
+
+    public ListenerHelper(Object owner) {
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java:79]: it is used in one of the skin draft PRs.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TestListenerHelper.java [hunk 11]
```diff
@@ -0,0 +1,465 @@
...
+        assertEquals(2, ct.get());
+
+        h.disconnect();
+
+        p.set("2");
+        assertEquals(2, ct.get());
+    }
+
+    @Test
+    public void testInvalidationListener_Callback() {
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/TestListenerHelper.java:278]: you found a bug in tests, thank you.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TestListenerHelper.java (hunk 8 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/TestListenerHelper.java:295]: another bug in tests.
+1

File: modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/IDisconnectable.java (hunk 9 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/IDisconnectable.java:34]: will decide later

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TestListenerHelper.java [hunk 12]
```diff
@@ -280,7 +280,7 @@ public void testInvalidationListener_Callback() {
         SimpleStringProperty p = new SimpleStringProperty();
         AtomicInteger ct = new AtomicInteger();
 
-        h.addInvalidationListener(p, (cur) -> ct.incrementAndGet());
+        h.addInvalidationListener(() -> ct.incrementAndGet(), p);
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/TestListenerHelper.java:283]: Now this is somewhat redundant with `testInvalidationListener_MultipleProperties`, since it tests the same flavor of method -- unlike `addChangeListener`, there are only two patterns, not three. Although showing that it works with a single property is possibly useful (the same could be said for the ChangeListener tests), so I don't mind if you prefer to keep it.

**Review by kevinrushforth — APPROVED** [review]
Looks OK, although I left another comment on the (now less so) redundant test.

**Comment by andy-goryachev-oracle** [comment]
oh no, all the approvals have been invalidated...

**Review by aghaisas — APPROVED** [review]
